spinnaker_components/common/input_filtering.c

```diff
@@ -0,0 +1,320 @@
+#include "nengo-common.h"
+#include "input_filtering.h"
+#include "common-impl.h"
+
+/* FILTER IMPLEMENTATIONS ****************************************************/
+/* None filter : output = input **********************************************/
+void _none_filter_step(uint32_t n_dims, value_t *input,
+                       value_t *output, void *params)
+{
+  use(params);
+
+  // The None filter just copies its input to the output
+  for (uint32_t d = 0; d < n_dims; d++)
+    output[d] = input[d];
+}
+
+void _none_filter_init(void *params, struct _if_filter *filter, uint32_t size)
+{
+  use(params);
+  use(size);
+
+  debug(">> None filter\n");
+
+  // We just need to set the function pointer for the step function.
+  filter->step = _none_filter_step;
+}
+
+/* 1st Order Low-Pass ********************************************************/
+struct _lowpass_state
+{
+  value_t a;  // exp(-dt / tau)
+  value_t b;  // 1 - a
+};
+
+void _lowpass_filter_step(uint32_t n_dims, value_t *input,
+                          value_t *output, void *pars)
+{
+  // Cast the params
+  struct _lowpass_state *params = (struct _lowpass_state *) pars;
+
+  // Apply the filter to every dimension (realised as a Direct Form I digital
+  // filter).
+  for (uint32_t d = 0; d < n_dims; d++)
+  {
+    output[d] *= params->a;
+    output[d] += input[d] * params->b;
```

Comment:
The ideal solution generated code wise would be for this to generate:

```
SMULL out_lo, out_hi, a, out 
SMLAL out_lo, out_hi, in, b
```

Which would result in an answer in a result in S33.30 spread across out_lo and out_hi. Then you could munge together the 32 bits from the middle of out_hi and out_lo and obtain the correct S16.15 result )